### DISCUSSION TITLE
How to stitch mesh A with mesh B after mesh A is cloned from another mesh?

### DISCUSSION THREAD

**Discussion by yuezhiying6** [idaholab/moose#23715] [Q&A General] [answered]
Hello all,
I want to merge two meshes into one mesh and found that the CombineGenerator cannot meet this command. In libMesh a function called stitch_meshes may be of use. However, there exist some errors for me to use this function. Part of my code is as follows:
```
// Here is a example
LibMeshInit init (argc, argv);
Mesh mesh_org(init.comm());
MeshTools::Generation::build_cube(mesh_org, 5,5,5,0,1,0,1,0,1,HEX8);
auto mesh2 = mesh_org.clone();
MeshTools::Modification::translate(*mesh2, 1, 0, 0);

auto mesh3 = mesh_org.clone();
MeshTools::Modification::translate(*mesh3, 2, 0, 0);

mesh2.stitch_meshes(*mesh3, 2, 4, TOLERANCE, true, true, false, false);
// Here is the error message

error: ‘__gnu_cxx::__alloc_traits<std::allocator<std::unique_ptr<libMesh::MeshBase> >, std::unique_ptr<libMesh::MeshBase> >::value_type’ {aka ‘class std::unique_ptr<libMesh::MeshBase>’} has no member named ‘stitch_meshes’
```
I would like to know how to correct or avoid this error. Any suggestions would be greatly appreciated.

**Comment by GiudGiud** [comment]
Hello

Did you try the StitchMeshGenerator instead?

For your error, mesh2 is a pointer (unique ptr returned by the clone())
so it should be 
```
mesh2->stitch_meshes(*mesh3, 2, 4, TOLERANCE, true, true, false, false);
```

A helpful resource for this is the doxygen btw
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1Mesh.html

Guillaume

**Reply by yuezhiying6** [reply]
Thanks for your reply. I just gave it a try as you suggested and another error occurred. 
class libMesh::MeshBase has no member named “stitch_meshes”

**Reply by GiudGiud** [reply]
MeshBase is probably too high in the inheritance chain

Try with an UnstructuredMesh class instead

**Reply by yuezhiying6** [reply]
Do you mean to set mesh_org as an UnstructuredMesh? I just tried as follows:
```
//Codes 
  LibMeshInit init (argc, argv);
  std::unique_ptr<UnstructuredMesh> mesh_org = std::make_unique<Mesh>(init.comm());
  MeshTools::Generation::build_cube(*mesh_org, 5, 5, 5, 0, 1, 0, 1, 0, 5, HEX8);
  auto mesh2 = mesh_org->clone();
  MeshTools::Modification::translate(*mesh2, 1, 0, 0);

  auto mesh3 = mesh_org->clone();
  MeshTools::Modification::translate(*mesh3, 1, 0, 0);
  mesh2.stitch_meshes(*mesh3, 2, 4, TOLERANCE, true, true, false, false);
  ExodusII_IO(*mesh2).write("final.e");
//Error 
error: ‘class std::unique_ptr<libMesh::MeshBase>’ has no member named ‘stitch_meshes’
   39 |   *mesh2.stitch_meshes(*mesh3, 2, 4, TOLERANCE, true, true, false, false);
```

I seem to have found a solution by translating mesh_org instead of mesh2, but this approach is somewhat inconvenient with multiple meshes.  If I can clone the mesh, it will be easier for me to work with the stitch_meshes.

Thank you again for your reply.

**Reply by GiudGiud** [reply, marked as answer]
mesh2 and mesh_org could be UnstructuredMeshes rather than Mesh
```
  std::unique_ptr<UnstructuredMesh> mesh_org = std::make_unique<Mesh>(init.comm());

```
should be
```
  std::unique_ptr<UnstructuredMesh> mesh_org = std::make_unique<UnstructuredMesh>(init.comm());
```

otherwise I think mesh2 is being created as a MeshBase rather than an unstructuredMesh

Anyhow you have the same problem with the pointer as before
`  mesh2.stitch_meshes(*mesh3, 2, 4, TOLERANCE, true, true, false, false);`
is not OK. mesh2 is a unique_ptr, so you can do this, you have to do:
`  mesh2->stitch_meshes(*mesh3, 2, 4, TOLERANCE, true, true, false, false);`

**Reply by yuezhiying6** [reply]
The return value of the function clone() is std::unique_ptr< MeshBase >. And it has no member function called stitched_meshes. So I think the type of mesh_org does not affect the type of mesh2. 
I found that the copy constructor of class Mesh may be of help. The following codes can run successfully.

//Codes
  LibMeshInit init (argc, argv);
  Mesh mesh_org(init.comm());
  MeshTools::Generation::build_cube(mesh_org, 5, 5, 5, 0, 1, 0, 1, 0, 5, HEX8);
  Mesh mesh2(mesh_org);
  MeshTools::Modification::translate(mesh2, 1, 0, 0);
  Mesh mesh3(mesh_org);
  MeshTools::Modification::translate(mesh3, 2, 0, 0);
  mesh2.stitch_meshes(mesh3, 2, 4, TOLERANCE, true, true, false, false);
  ExodusII_IO(mesh2).write("final.e");

Thank you very much for your help.
